pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/assignment/instance/InstanceReplicaGroupPartitionSelector.java

```diff
@@ -140,25 +132,118 @@ public void selectInstances(Map<Integer, List<InstanceConfig>> poolToInstanceCon
+
+          for (int partitionId = 0; partitionId < existingNumPartitions; partitionId++) {
+            List<String> existingInstances = _existingInstancePartitions.getInstances(partitionId, replicaGroupId);
+            replicaGroupIdToInstancesMap.computeIfAbsent(replicaGroupId, k -> new HashSet<>())
+                .addAll(existingInstances);
+          }
+        }
+
+        for (int replicaGroupId = 0; replicaGroupId < existingNumReplicaGroups; replicaGroupId++) {
+          Integer pool = replicaGroupIdToPoolMap.get(replicaGroupId);
+          if (pool == null) {
+            // Skip the replica group if it's no longer needed.
+            continue;
+          }
+          // Filter out instances that belong to other replica groups which should not be the candidate.
+          Set<String> candidateInstances = new LinkedHashSet<>(poolToCandidateInstancesMap.get(pool));
+          for (int otherReplicaGroupId = 0;
+              otherReplicaGroupId < existingNumReplicaGroups && otherReplicaGroupId < numReplicaGroups;
+              otherReplicaGroupId++) {
+            if (replicaGroupId != otherReplicaGroupId) {
+              candidateInstances.removeAll(replicaGroupIdToInstancesMap.get(otherReplicaGroupId));
+            }
+          }
+          Set<String> chosenCandidateInstances = new HashSet<>();
+          for (int partitionId = 0; partitionId < existingNumPartitions; partitionId++) {
+            List<String> existingInstances = _existingInstancePartitions.getInstances(partitionId, replicaGroupId);
+            List<String> instancesToSelect =
+                getInstancesWithMinimumMovement(numInstancesPerPartition, candidateInstances, existingInstances);
+            chosenCandidateInstances.addAll(instancesToSelect);
+            instancePartitions.setInstances(partitionId, replicaGroupId, instancesToSelect);
+          }
+          // Remove instances that are already been chosen.
+          poolToCandidateInstancesMap.get(pool).removeAll(chosenCandidateInstances);
+        }
+
+        // If the new number of replica groups is greater than the existing number of replica groups.
+        for (int replicaGroupId = existingNumReplicaGroups; replicaGroupId < numReplicaGroups; replicaGroupId++) {
+          int pool = replicaGroupIdToPoolMap.get(replicaGroupId);
+          Set<String> candidateInstances = new LinkedHashSet<>(poolToCandidateInstancesMap.get(pool));
+
+          Set<String> chosenCandidateInstances = new HashSet<>();
+          for (int partitionId = 0; partitionId < existingNumPartitions; partitionId++) {
+            List<String> existingInstances = _existingInstancePartitions.getInstances(partitionId, replicaGroupId);
+            if (existingInstances == null) {
+              existingInstances = Collections.emptyList();
+            }
+            List<String> instancesToSelect =
+                getInstancesWithMinimumMovement(numInstancesPerPartition, candidateInstances, existingInstances);
```

Comment:
I think this can cause uneven distribution if `numInstancesPerPartition` changes. We didn't track the number of partitions assigned to each server

Reply:
We should compute the maximum partitions per server, then use that as a limit to assign the partitions

Reply:
Good question! The `candidateInstances` here is a linked hash set. The purpose of that is to maintain the sequence as well as faster lookup.
For existing instances that are still alive, they will be popped out from the candidate list and then appended to the tail of the candidates, so that newly added instances has the chances to be picked up. For these existing alive instances, they will be validated by checking whether they exist in the candidates.
For newly added instances, the order of them remains unchanged, so that same instance can be picked up for all the partitions of a replica group. 
If it's a brand new RG, all the new instances in this candidate list will be picked up in round robin order.
I've written an analysis on all the possible cases in this doc, and unit tests in this PR cover all these scenarios. Please take a look:
https://docs.google.com/document/d/1_Fn-yNjt9Ih0SQiIqIhCEvE9BvYugvAhB4tgWH-VoWE/edit?usp=sharing

Reply:
Sure, added the comments to the code.